presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PhysicalCteOptimizer.java

```diff
@@ -0,0 +1,301 @@
+            // Create Table Metadata
+            PlanNode actualSource = node.getSource();
+            VariableReferenceExpression partitionVariable = actualSource.getOutputVariables()
+                    .get(getCtePartitionIndex(actualSource.getOutputVariables()));
+            List<Type> partitioningTypes = Arrays.asList(partitionVariable.getType());
+            String partitioningProviderCatalog = getPartitioningProviderCatalog(session);
+            // First column is taken as the partitioning column
+            Partitioning partitioning = Partitioning.create(
+                    metadata.getPartitioningHandleForExchange(session, partitioningProviderCatalog,
+                            getHashPartitionCount(session), partitioningTypes),
+                    Arrays.asList(partitionVariable));
+            BasePlanFragmenter.PartitioningVariableAssignments partitioningVariableAssignments
+                    = assignPartitioningVariables(variableAllocator, partitioning);
+            Map<VariableReferenceExpression, ColumnMetadata> variableToColumnMap =
+                    assignTemporaryTableColumnNames(actualSource.getOutputVariables(),
+                            partitioningVariableAssignments.getConstants().keySet());
+            List<VariableReferenceExpression> partitioningVariables = partitioningVariableAssignments.getVariables();
+            List<String> partitionColumns = partitioningVariables.stream()
+                    .map(variable -> variableToColumnMap.get(variable).getName())
+                    .collect(toImmutableList());
+            PartitioningMetadata partitioningMetadata = new PartitioningMetadata(partitioning.getHandle(), partitionColumns);
+
+            TableHandle temporaryTableHandle;
+            try {
+                temporaryTableHandle = metadata.createTemporaryTable(
+                        session,
+                        "hive",
+                        ImmutableList.copyOf(variableToColumnMap.values()),
+                        Optional.of(partitioningMetadata));
+                context.get().put(node.getCteName(),
+                        new PhysicalCteTransformerContext.TemporaryTableInfo(
+                                createTemporaryTableScan(
+                                        metadata,
+                                        session,
+                                        idAllocator,
+                                        node.getSourceLocation(),
+                                        temporaryTableHandle,
+                                        actualSource.getOutputVariables(),
+                                        variableToColumnMap,
+                                        partitioningMetadata), node.getOutputVariables()));
+            }
+            catch (PrestoException e) {
+                if (e.getErrorCode().equals(NOT_SUPPORTED.toErrorCode())) {
+                    throw new PrestoException(
+                            NOT_SUPPORTED,
```

Comment:
should we add this check to is isCteMaterializable?

Reply:
I think here its just a catch for the metadata.createTemporaryTable() which cannot be invoked in the parser since we do not have partition metadata and other info there